### ISSUE TITLE
[API Proposal]: Consider Equals method(s) on DefaultInterpolatedStringHandler to avoid temporary string allocations

### ISSUE THREAD

**Issue by BrennanConroy** [dotnet/runtime#110427]
### Background and motivation

Was just modifying some code and noticed it was creating a string from a number and adding it to other strings in an if statement just to compare to another string and then throw the just created string away. Started updating the code to remove the number allocation by using string interpolation, and then realized the interpolated string is still being allocated just to do a comparison and then thrown away. It seems a bit silly when we could do span comparisons to avoid the string allocation.

If this API was added, Roslyn would also want to update to generate code that makes use of the API as most people aren't going to be using `DefaultInterpolatedStringHandler` explicitly.

### API Proposal

```diff
namespace System.Runtime.CompilerServices;

public ref struct DefaultInterpolatedStringHandler
{
+    public bool Equals(ReadOnlySpan<char> other, StringComparison comparisonType);
-    internal void Clear();
+    public void Clear();
}
```

### API Usage

```csharp
var str1 = "somevalue";
var str2 = "some";
var val1 = 1;
if (str1 == $"{str2}{val1}")
{
}
```

Ideally Roslyn generated diff:
```diff
string text = "somevalue";
string value = "some";
int value2 = 1;
DefaultInterpolatedStringHandler defaultInterpolatedStringHandler = new DefaultInterpolatedStringHandler(0, 2);
defaultInterpolatedStringHandler.AppendFormatted(value);
defaultInterpolatedStringHandler.AppendFormatted(value2);
-bool flag = defaultInterpolatedStringHandler.ToStringAndClear() == text;
+bool flag = defaultInterpolatedStringHandler.Equals(text, StringComparison.InvariantCulture);
+defaultInterpolatedStringHandler.Clear();
```

### Alternative Designs

```diff
public ref struct DefaultInterpolatedStringHandler
{
+    public bool EqualsAndClear(ReadOnlySpan<char> other, StringComparison comparisonType);
}
```

### Risks

`DefaultInterpolatedStringHandler` uses an array pool internally and returns the array to the pool when the string is realized. Separating the usage of the interpolated string from the clearing makes the type more error prone, so the alternative design follows the current `ToStringAndClear` public method where it combines the action and clearing the array.

**Comment by huoyaoyuan** [comment]
`Equals` is just one operation that can reduce allocation. Exposing the formatted `Span` would benefit more scenarios.

**Comment by stephentoub** [comment]
> `Equals` is just one operation that can reduce allocation. Exposing the formatted `Span` would benefit more scenarios.

Yes, if we believe we have real need for this capability, we should reconsider https://github.com/dotnet/runtime/issues/55390 instead.

**Comment by BrennanConroy** [comment]
I probably should have put the alternative design as the primary one. Like https://github.com/dotnet/runtime/issues/55390 mentions, exposing more generic methods/properties makes the type more error prone. We could argue that since its primary use is by the compiler where we should be writing correct code, then making it a little more error prone is fine.

**Comment by BrennanConroy** [comment]
`Text` and `Clear` are now public and start to unblock the scenario here. Filed https://github.com/dotnet/roslyn/issues/77367 for roslyn to look into generating different code for equals comparisons.
